File: functionTests/getmodelTests/test1.R
.T_hdc_getTheModel = function(model, all2models = FALSE){
  # Function used to get the models from number or names
  
  model_in = model
  
  if(!is.vector(model)) stop("The argument 'model' must be a vector.")
  
  if(anyNA(model)) stop("The argument 'model' must not contain any NA.")
  
  ModelNames <- c("AKJBKQKDK", "AKBKQKDK", "ABKQKDK", "AKJBQKDK", "AKBQKDK", "ABQKDK", "AKJBKQKD", "AKBKQKD", "ABKQKD", "AKJBQKD", "AKBQKD", "ABQKD")
  
  model = toupper(model)
  
  if(length(model)==1 && model=="ALL"){
    if(all2models) model <- 1:14
    else return("ALL")
  }
  
  qui = which(model %in% 1:14)
  model[qui] = ModelNames[as.numeric(model[qui])]
  
  # We order the models properly
  qui = which(!model%in%ModelNames)
  if (length(qui)>0){
    if(length(qui)==1){
      msg = paste0("(e.g. ", model_in[qui], " is incorrect.)")
    } else {
      msg = paste0("(e.g. ", paste0(model_in[qui[1:2]], collapse=", or "), " are incorrect.)")
    }
    stop("Invalid model name ", msg)
  }
  
  # warning:
  if(max(table(model))>1) warning("The model vector, argument 'model', is made unique (repeated values are not tolerated).")
  
  mod_num <- c()
  for(i in 1:length(model)) mod_num[i] <- which(model[i]==ModelNames)
  mod_num <- sort(unique(mod_num))
  model <- ModelNames[mod_num]
  
  return(model)
}

a = c('a')
b = c(1)
c = c('a', 'b', 'c', 'd')
d = c("AKJBKQKDK", "a", "b")
e = c("AKJBKQKDK")

tests = list(a = a, b=b, c=c, d=d, e=e)

for(i in tests){
  try(print(.T_hdc_getTheModel(i)))
  
}

File: functionTests/getmodelTests/test1.py
import traceback
import numpy as np
def _T_hdc_getTheModel(model, all2models = False):
    
    model_in = model
    try:
        new_model = np.array(model,dtype='<U10')
        model = np.array(model)
    except:
        raise ValueError("Model needs to be an array or list")

    if(model.ndim > 1):
        raise ValueError("The argument 'model' must be 1-dimensional")
    
    if type(model[0]) != np.str_:
        if np.isnan(model):
            raise ValueError("The argument 'model' cannot contain any Nan")

    ModelNames = np.array(["AKJBKQKDK", "AKBKQKDK", "ABKQKDK", "AKJBQKDK", "AKBQKDK", "ABQKDK", "AKJBKQKD", "AKBKQKD", "ABKQKD", "AKJBQKD", "AKBQKD", "ABQKD"])
    if type(model[0]) == np.str_:
        model = [np.char.upper(m) for m in model]

    if len(model) == 1 and model[0] == "ALL":
        if all2models:
            model = np.arange(0, 13)
        else:
            return "ALL"
        
    if type(model[0]) == np.int_:
        qui = np.nonzero(model in np.arange(0, 13))
        if len(qui) > 0:
            new_model[qui] = ModelNames[model[qui]]
        #print(new_model)
        
    qui = np.nonzero( np.invert(np.isin(new_model, ModelNames)))[0]
    if len(qui) > 0:
        if len(qui) == 1:
            msg = f'(e.g. {model_in[qui[0]]} is incorrect.)'

        else:
            msg = f'(e.g. {model_in[qui[0]]} or {model_in[qui[1]]} are incorrect.)'

        raise ValueError("Invalid model name " + msg)
    
    if np.max(np.unique(model, return_counts=True)[1]) > 1:
        raise ValueError("Values in 'model' argument should be unique.")

    mod_num = []
    for i in range(len(new_model)):
        mod_num.append(np.nonzero(new_model[i] == ModelNames)[0])
    mod_num = np.sort(np.unique(mod_num))
    model = ModelNames[mod_num]

    return new_model

a = ["a"]
b = [1]
c = ["a", "b", "c", "d"]
d = ["AKJBKQKDK", "a", "b"]
e = ["AKJBKQKDK"]

tests = [a,b,c,d,e]

for model in tests:
    try:
        print(_T_hdc_getTheModel(model))
    except Exception as e:
        print("error " + str(e))
        #print(traceback.print_exc())
        